src/Cake.Issues/FileLinking/AzureDevOpsFileLinkSettingsBuilder.cs

```diff
@@ -0,0 +1,89 @@
+        internal AzureDevOpsFileLinkSettingsBuilder(Uri repositoryUrl)
+        {
+            repositoryUrl.NotNull(nameof(repositoryUrl));
+
+            this.repositoryUrl = repositoryUrl;
+        }
+
+        /// <summary>
+        /// Returns settings to files on AzureDevOps on a specific branch.
+        /// </summary>
+        /// <param name="branchName">Name of the branch.</param>
+        /// <returns>Class for setting additional settings.</returns>
+        public FileLinkOptionalSettingsBuilder Branch(string branchName)
+        {
+            branchName.NotNullOrWhiteSpace(nameof(branchName));
+
+            return
+                new FileLinkOptionalSettingsBuilder(
+                    (issue, values) =>
+                    {
+                        issue.NotNull(nameof(issue));
+                        values.NotNull(nameof(values));
+
+                        var fileLinkPattern =
+                            this.GetFileLinkPattern(issue, values, "GB", branchName);
+                        return new Uri(fileLinkPattern.ReplaceIssuePattern(issue));
+                    });
+        }
+
+        /// <summary>
+        /// Returns settings to files on AzureDevOps on a specific commit.
+        /// </summary>
+        /// <param name="commitId">The commit id on which the file linking will be based on.</param>
+        /// <returns>Class for setting additional settings.</returns>
+        public FileLinkOptionalSettingsBuilder Commit(string commitId)
+        {
+            commitId.NotNullOrWhiteSpace(nameof(commitId));
+
+            return
+                new FileLinkOptionalSettingsBuilder(
+                    (issue, values) =>
+                    {
+                        issue.NotNull(nameof(issue));
+                        values.NotNull(nameof(values));
+
+                        var fileLinkPattern =
+                            this.GetFileLinkPattern(issue, values, "GC", commitId);
+                        return new Uri(fileLinkPattern.ReplaceIssuePattern(issue));
+                    });
+        }
+
+        private string GetFileLinkPattern(
+            IIssue issue,
+            IDictionary<string, string> values,
+            string versionType,
+            string version)
+        {
+            issue.NotNull(nameof(issue));
+            values.NotNull(nameof(values));
+
+            var linePattern = issue.Line.HasValue ? "{Line}" : "1";
+            return
+                this.repositoryUrl.ToString().TrimEnd('/') +
+                "?path=" + values["rootPath"].WithEnding("/") + "{FilePath}" +
+                "&version=" + versionType + version +
+                "&line=" + linePattern +
+                "&lineEnd=" + (issue.EndLine.HasValue ? "{EndLine}" : linePattern) +
+                "&lineStartColumn=" + (issue.Column.HasValue ? "{Column}" : "1") +
+                "&lineEndColumn=" + (issue.EndColumn.HasValue ? "{EndColumn}" : "1");
```

Comment:
This does make no sense to me. I mean, you either set a placeholder or set a default value. But when deciding what to do, you already access the property. So why do you not insert `EndLine`, `Column` and `EndColumn` directly into the pattern, when you already access them for the check?

Reply:
Sorry, was confused, the patterns are resolved [here](https://github.com/cake-contrib/Cake.Issues/pull/199/files#diff-cbfc6a3f46aa78fddd74409fb2312cc9R43), but `rootPath` needs to be read differently.